lib/membrane_opus/parser.ex

```diff
@@ -1,79 +1,254 @@
 defmodule Membrane.Opus.Parser do
   @moduledoc """
-  Parses self-delimiting Opus stream.
+  Parses a raw incoming Opus stream and adds caps information, as well as metadata.
 
-  The self-delimiting version of Opus packet contains all the information about frame sizes
-  needed to separate the packet from byte stream and decode it. This parser drops the last frame size
-  from each packet header, converting packets to the basic version. The remaining frame size needs
-  to be derived from entire packet size.
-  See https://tools.ietf.org/html/rfc6716#appendix-B for details.
+  Adds the following metadata:
+
+  frame_size :: 2.5 | 5 | 10 | 20 | 40 | 60
+    Number of milliseconds per frame encoded in this packet.
+
+  frame_lengths :: [non_neg_integer()]
+    Ordered list of bytes used to encode each frame in this packet.
+    The length of this list is also the number of frames encoded in this packet.
+    However, a value of 0 indicates a dropped frame and should be accounted
+    for when using this field.
   """
+
   use Membrane.Filter
 
-  alias Membrane.{Buffer, Opus, RemoteStream}
-  alias Membrane.Opus.PacketUtils
+  alias Membrane.{Buffer, Opus}
+
+  def_options self_delimit?: [
+                spec: boolean(),
+                default: false,
+                description: """
+                If true, will mutate the output to self-delimit the Opus input.
 
-  def_input_pad :input,
-    demand_unit: :buffers,
-    caps: [{Opus, self_delimiting?: true}, {RemoteStream, content_format: one_of([Opus, nil])}]
+                See https://tools.ietf.org/html/rfc6716#appendix-B for details.
+                """
+              ]
 
-  def_output_pad :output, caps: {Opus, self_delimiting?: false}
+  def_input_pad :input, demand_unit: :buffers, caps: :any
+  def_output_pad :output, caps: Opus
 
   @impl true
-  def handle_init(_opts) do
-    {:ok, %{partial_packet: <<>>, timestamp: 0}}
+  def handle_init(%__MODULE__{} = options) do
+    {:ok, options |> Map.from_struct()}
   end
 
   @impl true
-  def handle_caps(:input, _caps, _ctx, state) do
-    {:ok, state}
+  def handle_demand(:output, bufs, :buffers, _ctx, state) do
+    {{:ok, demand: {:input, bufs}}, state}
   end
 
   @impl true
-  def handle_demand(:output, _size, _unit, _ctx, state) do
-    {{:ok, demand: {:input, 1}}, state}
+  def handle_process(:input, %Buffer{payload: data}, _ctx, state) do
```

Comment:
it doesn't. in order to do this you would need another option to say that the input is indeed self-delimitting. as i mentioned, i don't think there is any way to determine that a packet _IS_ self-delimiting just from looking at it.

Reply:
It seems you're right, such option would have to be added. But we're going to need that to make packets taken from OGG non-self-delimiting, what is required to decode them.

Reply:
can we add this option/feature later?

Reply:
Sure, but we'll need to somehow leave the current parser there until this is added

Reply:
oh i see - you're not saying that you want to guard against double-self-delimiting, instead you want to _undelimit_. can you confirm? if so i'm thinking of basically just adding one more option: unself_delimit. but what do you think?

Reply:
or perhaps, a single option:

`delimitation: :delimit | :undelimit | :none` with a default of `:none`. then the following are legal:

input is delimitted, you can choose :undelimit or :none
input is not delimitted, you can choose :delimit or :none

other combos are errors. should it even raise at that point?

Reply:
> or perhaps, a single option:
delimitation: :delimit | :undelimit | :none with a default of :none

good idea, you may also choose `keep` instead of `none`

> other combos are errors. should it even raise at that point?

I think it should